src/analysis/inner_product_space/gram_schmidt_ortho.lean

```diff
@@ -0,0 +1,213 @@
+
+## TODO
+  Construct a version with an orthonormal basis from the Gram-Schmidt process.
+-/
+
+open_locale big_operators
+
+variables (𝕜 : Type*) {E : Type*} [is_R_or_C 𝕜] [inner_product_space 𝕜 E]
+
+local notation `⟪`x`, `y`⟫` := @inner 𝕜 _ _ x y
+
+/-- Gram-Schmidt process -/
+noncomputable def gram_schmidt (f : ℕ → E) : ℕ → E
+| n := f n - ∑ i : fin n, orthogonal_projection (𝕜 ∙ gram_schmidt i) (f n)
+using_well_founded {dec_tac := `[exact i.prop]}
+
+/-- 'gram_schmidt_def' turns the sum over `fin n` into a sum over `ℕ`. -/
+lemma gram_schmidt_def (f : ℕ → E) (n : ℕ) :
+  gram_schmidt 𝕜 f n = f n - ∑ i in finset.range n,
+    orthogonal_projection (𝕜 ∙ gram_schmidt 𝕜 f i) (f n) :=
+begin
+  rw gram_schmidt,
+  congr' 1,
+  exact fin.sum_univ_eq_sum_range (λ i,
+    (orthogonal_projection (𝕜 ∙ gram_schmidt 𝕜 f i) (f n) : E)) n,
+end
+
+lemma gram_schmidt_def' (f : ℕ → E) (n : ℕ):
+  f n = gram_schmidt 𝕜 f n + ∑ i in finset.range n,
+    orthogonal_projection (𝕜 ∙ gram_schmidt 𝕜 f i) (f n) :=
+by simp only [gram_schmidt_def, sub_add_cancel]
+
+@[simp] lemma gram_schmidt_zero (f : ℕ → E) :
+  gram_schmidt 𝕜 f 0 = f 0 :=
+by simp only [gram_schmidt, fintype.univ_of_is_empty, finset.sum_empty, sub_zero]
+
+/-- **Gram-Schmidt Orthogonalisation**
+Gram-Schmidt process produces an orthogonal system of vectors. -/
+theorem gram_schmidt_orthogonal (f : ℕ → E) (a b : ℕ) (h₀ : a ≠ b) :
+  ⟪gram_schmidt 𝕜 f a, gram_schmidt 𝕜 f b⟫ = 0 :=
+begin
+  suffices : ∀ a b : ℕ, a < b → ⟪gram_schmidt 𝕜 f a, gram_schmidt 𝕜 f b⟫ = 0,
+  { cases h₀.lt_or_lt with ha hb,
+    { exact this _ _ ha, },
+    { rw inner_eq_zero_sym,
+      exact this _ _ hb, }, },
+  clear h₀ a b,
+  intros a b h₀,
+  obtain ⟨c, hbc⟩ : ∃ c, b ≤ c := ⟨b, le_rfl⟩,
+  induction c using nat.strong_induction_on with c hc generalizing a b,
+  rcases hbc.lt_or_eq with (hbc | rfl),
+  { exact hc b hbc a b h₀ le_rfl, },
+  simp only [gram_schmidt_def 𝕜 f b, inner_sub_right, inner_sum,
+    orthogonal_projection_singleton, inner_smul_right],
+  rw finset.sum_eq_single_of_mem a (finset.mem_range.mpr h₀),
+  { by_cases h : gram_schmidt 𝕜 f a = 0,
+    { simp only [h, inner_zero_left, zero_div, zero_mul, sub_zero], },
+    { rw [← inner_self_eq_norm_sq_to_K, div_mul_cancel, sub_self],
+      rwa [ne.def, inner_self_eq_zero], }, },
+  intros i hi hia,
+  simp only [mul_eq_zero, div_eq_zero_iff, inner_self_eq_zero],
+  right,
+  rw finset.mem_range at hi,
+  cases hia.lt_or_lt with hia₁ hia₂,
+  { rw inner_eq_zero_sym,
+    exact hc a h₀ i a hia₁ le_rfl, },
+  { exact hc i hi a i hia₂ le_rfl, },
+end
+
+open submodule set
+
+/-- Gram-Schmidt process preserves span -/
+lemma span_gram_schmidt (f : ℕ → E) (c : ℕ) :
+  span 𝕜 (gram_schmidt 𝕜 f '' Iic c) = span 𝕜 (f '' Iic c) :=
+begin
+  induction c with c hc,
+  { simp only [Iic, gram_schmidt_zero, le_zero_iff, set_of_eq_eq_singleton, image_singleton], },
+  have h : Iic c.succ = insert c.succ (Iic c),
```

Comment:
Just a reminder that before merging, we should either wait for @YaelDillies' PR that adds this lemma, or extract it to a suitable file as part of this PR.

Reply:
I'm PRing it now.

Reply:
Thank you! I have added it on dependent list